src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -977,11 +977,65 @@
                                pids
                                values)))))
 
+(defn find-certname-id-and-hash [certname]
+  (first (jdbc/query-to-vec [(format "SELECT id, %s as package_hash FROM certnames WHERE certname=?"
+                                     (sutils/sql-hash-as-str "package_hash"))
+                             certname])))
+
+(defn create-package-map [certname-id]
+  (jdbc/call-with-query-rows ["SELECT id, name, version, provider FROM package_inventory WHERE certname_id=?"
+                              certname-id]
+                             (fn [rows]
+                               (->> rows
+                                    (map (fn [{id :id package_name :name version :version provider :provider}]
+                                           [[package_name version provider] id]))
+                                    (into {})))))
+
+(s/defn update-packages
+  "Compares `inventory` to the stored package inventory for
+  `certname`. Differences will result in updates to the database"
+  [certname
+   inventory :- [facts/package-tuple]]
+  (let [{certname-id :id package_hash :package_hash} (find-certname-id-and-hash certname)
+        new-package-hash (shash/package-similarity-hash inventory)]
+
+    (when-not (= new-package-hash package_hash)
+      (let [old-packages (create-package-map certname-id)
+            [to-add to-remove _] (clojure.data/diff (set inventory) (set (keys old-packages)))
+            to-remove-ids (map old-packages to-remove)]
+        (jdbc/update! :certnames
+                      {:package_hash (sutils/munge-hash-for-storage new-package-hash)}
+                      ["id=?" certname-id])
+        (when (seq to-add)
+          (jdbc/insert-multi! :package_inventory
+                              ["certname_id" "name" "version" "provider"]
+                              (map #(cons certname-id %) to-add)))
+        (when (seq to-remove)
+          (jdbc/delete! :package_inventory
+                        (cons (str "id " (jdbc/in-clause to-remove-ids))
```

Comment:
use an array param here, to be kinder to the connection pool's prepared statement cache